plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -322,6 +332,39 @@ private List<FileScanTaskWithDomain> processFileScanTask(TupleDomain<IcebergColu
             return ImmutableList.of();
         }
 
+        if (!recordScannedFiles || scannedFilesByPartition == null) {
+            return ImmutableList.of(fileScanTaskWithDomain);
+        }
+
+        // Assess if the partition that wholeFileTask belongs to should be included for OPTIMIZE
+        // If file was partitioned under an old spec, OPTIMIZE may be able to merge it with another file under new partitioning spec
+        // We don't know which partition of new spec this file belongs to, so we include all files in OPTIMIZE
+        if (currentSpecId != wholeFileTask.spec().specId()) {
+            Stream<FileScanTaskWithDomain> allQueuedTasks = scannedFilesByPartition.values().stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get);
+            scannedFilesByPartition = null;
+            return Stream.concat(allQueuedTasks, Stream.of(fileScanTaskWithDomain)).collect(toImmutableList());
+        }
+        Types.StructType structType = wholeFileTask.spec().partitionType();
+        StructLikeWrapper partitionWrapper = StructLikeWrapper.forType(structType).set(wholeFileTask.file().partition());
+        StructLikeWrapperWithFieldIdToIndex structLikeWrapperWithFieldIdToIndex = new StructLikeWrapperWithFieldIdToIndex(partitionWrapper, structType);
+        Optional<FileScanTaskWithDomain> alreadyQueuedFileTask = scannedFilesByPartition.get(structLikeWrapperWithFieldIdToIndex);
+        if (alreadyQueuedFileTask != null) {
+            // Optional.empty() is a marker for partitions where we've seen enough files to avoid skipping them from OPTIMIZE
```

Comment:
Consider replacing with this comment

https://github.com/trinodb/trino/blob/cabe3aefa1e2b5ddf1efe7cd1a2a21eebe6592e9/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java#L254

Reply:
Algorithm similar to what has been done in https://github.com/trinodb/trino/commit/3aea15e26d51f555ea8a78cc073a80bb1564f9ef

Reply:
I'm not working with a stream here, the current comment feels easier to read to me. We also have a special case where the first file encountered in the partition has deletes, then we put the marker straightaway and don't wait to see 2 files.